src/utils/block.ts

```diff
@@ -0,0 +1,211 @@
+import { v4 as makeUUID } from 'uuid';
+import * as pug from 'pug';
+import EventBus from './event-bus';
+
+export default abstract class Block<T extends object = {}> {
+  public static EVENTS = {
+    INIT: 'init',
+    FLOW_CDM: 'flow:component-did-mount',
+    FLOW_CDU: 'flow:component-did-update',
+    FLOW_RENDER: 'flow:render',
+  };
+
+  private _element: HTMLElement | null = null;
+
+  private _meta: { tagName: string } = { tagName: 'div' };
+
+  private _events: Record<string, (e: Event) => void> = {};
+
+  public id: string | null = null;
+
+  public props: T;
+
+  public children: Record<string, Block> = {};
+
+  public eventBus: EventBus;
+
+  public state: Record<string, unknown> = this._makeProxyObj({});
+
+  constructor(tagName = 'div', props: T = <T>{}) {
+    this._meta = { tagName };
+
+    this.id = makeUUID();
+    this.props = this._makeProxyObj(props);
+
+    const eventBus = new EventBus();
+    this.eventBus = eventBus;
```

Comment:
```suggestion
    this.eventBus = new EventBus();
```

как будто так получится проще)